pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/BaseResource.java

```diff
@@ -57,25 +61,53 @@ protected BaseResource(Authentication auth) {
         this.auth = auth;
     }
 
-    public Builder request(final WebTarget target) throws PulsarAdminException {
+    // do the authentication stage, and once authentication completed return a Builder
+    public CompletableFuture<Builder> request(final WebTarget target) throws PulsarAdminException {
```

Comment:
I think you shouldn't change the signature of this method. 

```
public Builder request(final WebTarget target) throws PulsarAdminException {
     try {
        return requestAsync(target).get();
     }  catch (Exception e) {
        if (e.getCause() instanceof PulsarAdminException) {
            throw (PulsarAdminException) (e.getCause());
        } else {
            throw new PulsarAdminException(..., e);
        }
     }
}

public CompletableFuture<Builder> requestAsync(final WebTarget target) {
    // ... implement the async method
}
```

so that you don't need to change the places calling `request(target)`. what you only need to change is those `async*Request` methods.